src/v/cluster/cluster_discovery.cc

```diff
@@ -80,6 +83,25 @@ ss::future<bool> cluster_discovery::is_cluster_founder() {
     if (_is_cluster_founder.has_value()) {
         co_return *_is_cluster_founder;
     }
+    // If there's anything in the controller directory, assume this node has
+    // previously joined a cluster.
+    auto controller_ntp_cfg = storage::ntp_config(
+      model::controller_ntp, config::node().data_directory().as_sstring());
+    const auto controller_dir = controller_ntp_cfg.work_directory();
+    auto controller_dir_exists = co_await ss::file_exists(controller_dir);
+    if (controller_dir_exists) {
+        const auto controller_empty = co_await directory_walker::empty(
+          std::filesystem::path(controller_dir));
+        if (!controller_empty) {
+            vlog(
+              clusterlog.info,
+              "Controller directory {} not empty; assuming existing cluster "
+              "exists",
+              controller_dir);
+            _is_cluster_founder = false;
+            co_return *_is_cluster_founder;
+        }
+    }
```

Comment:
I think this can be a static function in `storage::api` so that the knowledge about directory structure stays encapsulated there

Reply:
To your point about encapsulating directory structure, I think the real issue is that I didn't want to introduce a dependency on `log` or `log_manager` here, as this happens before most things are initialized. I'm not sure where else `is_log_dir_empty()` would be used, so I'm leaning away from putting this anywhere else unless we are expecting to use this elsewhere.

If anything, I think some `storage::log_dir_for_topic()` might be handy. I haven't looked around storage enough to see where that might be used but I imagine adopting such a primitive would involve a potentially broad refactor though, so I'll probably punt on it for now.

Reply:
If you mean dependency as a reference to a `storage::api` object/service, that won't be  needed if `is_log_dir_empty()` is a static method - and there's no reason for it not to be static I think.